# The DataFrame and Series

The DataFrame and Series are the two primary objects when using pandas to analyze data. In this chapter, we will learn how to read in data into a DataFrame and understand its components. We will also learn how to select a single column of data as a Series and examine its components.

## Reading external data with pandas

The one thing you need for data analysis is **data**. If you do not have any data, then you won't be able to use pandas to analyze it. This book contains many data sets stored externally in the `data` directory one level above where this notebook resides. Most of these datasets are stored as comma-separated value (**CSV**) files. These CSVs are human-readable and separate each individual piece of data with a comma. The comma is referred to as the **delimiter**. Despite its name, CSVs can use other one-character delimiters besides commas such as tabs, semi-colons, or others. 


### City of Chicago bike rides

We begin our data analysis adventure with a dataset on public bike rides from the city of Chicago. The data is contained in the `bikes.csv` file. There are about 50,000 recorded rides from 2013 through 2017. Each row of the dataset represents a single ride from a single person using the city's public bike stations. There are 19 columns of data containing information on gender, start time, trip duration, bike station name, temperature, wind speed, and more. Let's print out the first three lines of the `bikes.csv` file using Python's built-in capabilities for reading files. This does not use pandas. Take note of the commas separating each value on each line.

In [1]:
with open('../data/bikes.csv') as f:
    for i in range(3):
        print(f.readline())     

gender,starttime,stoptime,tripduration,from_station_name,start_capacity,to_station_name,end_capacity,temperature,wind_speed,events

Male,2013-06-28 19:01:00,2013-06-28 19:17:00,993,Lake Shore Dr & Monroe St,11.0,Michigan Ave & Oak St,15.0,73.9,12.7,mostlycloudy

Male,2013-06-28 22:53:00,2013-06-28 23:03:00,623,Clinton St & Washington Blvd,31.0,Wells St & Walton St,19.0,69.1,6.9,partlycloudy



### Understanding the file location

Above, the string `../data/bikes.csv` was used to represent the file location of the data. This location is relative to the directory where this notebook resides on your machine. Let's cover every part of this string to ensure we understand what it means.

The file location string begins with two dots, `..`. This translates as "move one level above the current working directory" to the **Jupyter Notebooks** directory. Appearing next in the string is `/data`, which translates as 'move down into the `data` directory. 

Note that the forward slash was written to separate the directories. Both macOS and Linux operating systems use this forward slash to separate directories and files from one another. On the other hand, the Windows operating system uses the backslash. Fortunately, we can always use a forward slash regardless of our operating system, as Python will automatically handle the file location string for us.

The string ends with `/bikes.csv` which translates as 'reference the filename `bikes.csv`. In summary, the file location `../data/bikes.csv` represents a relative location to where the dataset resides.

### Import pandas

To use the pandas library, we need to import it into our namespace. By convention, pandas is imported and aliased to the name `pd`. After running the import statement below, we will have access to all pandas objects with variable name `pd`. It is possible to use any other valid variable name as an alias, but it's best to use `pd` as the official documentation uses it along with most everyone else.

In [2]:
import pandas as pd

### Reading a CSV with the `read_csv` function

pandas provides the `read_csv` function to read in CSV files into a pandas DataFrame. The first argument passed to `read_csv` needs to be the location of the file relative to the current directory as a string, which is the same string from above. Let's call this function to read in our data.

In [3]:
bikes = pd.read_csv('../data/bikes.csv')

bikes

gender            starttime             stoptime  tripduration  \
0        Male  2013-06-28 19:01:00  2013-06-28 19:17:00           993   
1        Male  2013-06-28 22:53:00  2013-06-28 23:03:00           623   
2        Male  2013-06-30 14:43:00  2013-06-30 15:01:00          1040   
3        Male  2013-07-01 10:05:00  2013-07-01 10:16:00           667   
4        Male  2013-07-01 11:16:00  2013-07-01 11:18:00           130   
...       ...                  ...                  ...           ...   
50084    Male  2017-12-30 13:07:00  2017-12-30 13:34:00          1625   
50085    Male  2017-12-30 13:34:00  2017-12-30 13:44:00           585   
50086    Male  2017-12-30 13:34:00  2017-12-30 13:48:00           824   
50087  Female  2017-12-31 09:30:00  2017-12-31 09:33:00           178   
50088    Male  2017-12-31 15:22:00  2017-12-31 15:26:00           214   

                  from_station_name  start_capacity  \
0         Lake Shore Dr & Monroe St            11.0   
1      Clinton St & Washington Blvd            31.0   
2      Sheffield Ave & Kingsbury St            15.0   
3           Carpenter St & Huron St            19.0   
4            Damen Ave & Pierce Ave            19.0   
...                             ...             ...   
50084         State St & Pearson St            27.0   
50085      Halsted St & 35th St (*)            16.0   
50086      Kingsbury St & Kinzie St            31.0   
50087          Clinton St & Lake St            23.0   
50088    Clarendon Ave & Leland Ave            15.0   

                     to_station_name  end_capacity  temperature  wind_speed  \
0              Michigan Ave & Oak St          15.0         73.9        12.7   
1               Wells St & Walton St          19.0         69.1         6.9   
2            Dearborn St & Monroe St          23.0         73.0        16.1   
3             Clark St & Randolph St          31.0         72.0        16.1   
4             Damen Ave & Pierce Ave          19.0         73.0        17.3   
...                              ...           ...          ...         ...   
50084              Clark St & Elm St          27.0          5.0        16.1   
50085            Union Ave & Root St          11.0          5.0        16.1   
50086  Halsted St & Blackhawk St (*)          20.0          5.0        16.1   
50087       Kingsbury St & Kinzie St          31.0          7.0        11.5   
50088     Clifton Ave & Lawrence Ave          15.0         10.9        15.0   

             events  
0      mostlycloudy  
1      partlycloudy  
2      mostlycloudy  
3      mostlycloudy  
4      partlycloudy  
...             ...  
50084  partlycloudy  
50085  partlycloudy  
50086  partlycloudy  
50087  partlycloudy  
50088  partlycloudy  

[50089 rows x 11 columns]

### Display DataFrame in Jupyter Notebook

We assigned the output from the `read_csv` function to the `bikes` variable name which now refers to a DataFrame object. To visually display the DataFrame, place the variable name as the last line in a code cell. By default, pandas outputs the first and last 5 rows and first and last 10 columns. If there are less than 60 total rows, it displays all rows. We cover how to change these display options in an upcoming chapter.

### `head` and `tail` methods

A very useful and simple method is `head`, which returns the first 5 rows of the DataFrame by default. This avoids long default output and is something I highly recommend when doing data analysis within a notebook. The `tail` method returns the last 5 rows by default. There will only be a few instances in the book where the `head` method is not used, as displaying up to 60 rows is far too many and will take up a lot of space on a screen or page.

In [4]:
bikes.head()

gender            starttime             stoptime  tripduration  \
0   Male  2013-06-28 19:01:00  2013-06-28 19:17:00           993   
1   Male  2013-06-28 22:53:00  2013-06-28 23:03:00           623   
2   Male  2013-06-30 14:43:00  2013-06-30 15:01:00          1040   
3   Male  2013-07-01 10:05:00  2013-07-01 10:16:00           667   
4   Male  2013-07-01 11:16:00  2013-07-01 11:18:00           130   

              from_station_name  start_capacity          to_station_name  \
0     Lake Shore Dr & Monroe St            11.0    Michigan Ave & Oak St   
1  Clinton St & Washington Blvd            31.0     Wells St & Walton St   
2  Sheffield Ave & Kingsbury St            15.0  Dearborn St & Monroe St   
3       Carpenter St & Huron St            19.0   Clark St & Randolph St   
4        Damen Ave & Pierce Ave            19.0   Damen Ave & Pierce Ave   

   end_capacity  temperature  wind_speed        events  
0          15.0         73.9        12.7  mostlycloudy  
1          19.0         69.1         6.9  partlycloudy  
2          23.0         73.0        16.1  mostlycloudy  
3          31.0         72.0        16.1  mostlycloudy  
4          19.0         73.0        17.3  partlycloudy

The last five rows of the DataFrame may be displayed with the `tail` method.

In [5]:
bikes.tail()

gender            starttime             stoptime  tripduration  \
50084    Male  2017-12-30 13:07:00  2017-12-30 13:34:00          1625   
50085    Male  2017-12-30 13:34:00  2017-12-30 13:44:00           585   
50086    Male  2017-12-30 13:34:00  2017-12-30 13:48:00           824   
50087  Female  2017-12-31 09:30:00  2017-12-31 09:33:00           178   
50088    Male  2017-12-31 15:22:00  2017-12-31 15:26:00           214   

                from_station_name  start_capacity  \
50084       State St & Pearson St            27.0   
50085    Halsted St & 35th St (*)            16.0   
50086    Kingsbury St & Kinzie St            31.0   
50087        Clinton St & Lake St            23.0   
50088  Clarendon Ave & Leland Ave            15.0   

                     to_station_name  end_capacity  temperature  wind_speed  \
50084              Clark St & Elm St          27.0          5.0        16.1   
50085            Union Ave & Root St          11.0          5.0        16.1   
50086  Halsted St & Blackhawk St (*)          20.0          5.0        16.1   
50087       Kingsbury St & Kinzie St          31.0          7.0        11.5   
50088     Clifton Ave & Lawrence Ave          15.0         10.9        15.0   

             events  
50084  partlycloudy  
50085  partlycloudy  
50086  partlycloudy  
50087  partlycloudy  
50088  partlycloudy

### First and last `n` rows
Both the `head` and `tail` methods accept a single integer parameter `n` controlling the number of rows returned. Here, we output the first three rows.

In [6]:
bikes.head(3)

gender            starttime             stoptime  tripduration  \
0   Male  2013-06-28 19:01:00  2013-06-28 19:17:00           993   
1   Male  2013-06-28 22:53:00  2013-06-28 23:03:00           623   
2   Male  2013-06-30 14:43:00  2013-06-30 15:01:00          1040   

              from_station_name  start_capacity          to_station_name  \
0     Lake Shore Dr & Monroe St            11.0    Michigan Ave & Oak St   
1  Clinton St & Washington Blvd            31.0     Wells St & Walton St   
2  Sheffield Ave & Kingsbury St            15.0  Dearborn St & Monroe St   

   end_capacity  temperature  wind_speed        events  
0          15.0         73.9        12.7  mostlycloudy  
1          19.0         69.1         6.9  partlycloudy  
2          23.0         73.0        16.1  mostlycloudy

## Components of a DataFrame

The DataFrame is composed of three separate components - the **columns**, the **index**, and the **data**. These terms will be used throughout the book and understanding them is vital to your ability to use pandas. Take a look at the following graphic of our `bikes` DataFrame stylized to put emphasis on each component.

![1]

### The columns

The columns provide a **label** for each column and are always displayed in **bold** font above the data. A column is a single vertical sequence of data. In the above DataFrame, the column name `tripduration` references all the values in that column (993, 623, 1040, etc...).

The columns are also referred to as the **column names** or the **column labels** with individual values referred to as a **column name** or **column label**.

Most DataFrames, like the one above, use strings for column names, but it is possible that they can be other types such as integers. The column names are not required to be unique, though having duplicate columns would be bad practice, as it's vital to be able to uniquely identify each column.

### The index

The index provides a **label** for each row and is always displayed to the left of the data. A row is a single horizontal sequence of data. For instance, the index label **3** references all the values in its row (12907, Subscriber, Male, etc...)

The index is also referred to as the **index names/labels** or the **row names/labels** with the individual values referred to as a(n) **index name/label** or **row name/label**.

In the above DataFrame, the index is simply a sequence of integers beginning at 0. The values in the index are not limited to integers. Strings are a common type that are used in the index and make for more descriptive labels.

Surprisingly, values in the index are not required to be unique. In fact, all of the index values can be the same. A row label does not guarantee a one-to-one mapping to one specific row.

### The data

The actual data is to the right of the index and below the columns and is displayed with normal font. The data is also referred to as the **values**. The data represents all the values for all the columns. It is important to note that the index and the columns are NOT part of the data. They are separate objects that act as **labels** for either rows or columns.

### The Axes

The index and columns are known collectively as the **axes**, each representing a single **axis** of the two-dimensional DataFrame. pandas uses the integer **0** to reference the index and **1** for the columns.

[1]: images/df_components.png

In [7]:
bikes

gender            starttime             stoptime  tripduration  \
0        Male  2013-06-28 19:01:00  2013-06-28 19:17:00           993   
1        Male  2013-06-28 22:53:00  2013-06-28 23:03:00           623   
2        Male  2013-06-30 14:43:00  2013-06-30 15:01:00          1040   
3        Male  2013-07-01 10:05:00  2013-07-01 10:16:00           667   
4        Male  2013-07-01 11:16:00  2013-07-01 11:18:00           130   
...       ...                  ...                  ...           ...   
50084    Male  2017-12-30 13:07:00  2017-12-30 13:34:00          1625   
50085    Male  2017-12-30 13:34:00  2017-12-30 13:44:00           585   
50086    Male  2017-12-30 13:34:00  2017-12-30 13:48:00           824   
50087  Female  2017-12-31 09:30:00  2017-12-31 09:33:00           178   
50088    Male  2017-12-31 15:22:00  2017-12-31 15:26:00           214   

                  from_station_name  start_capacity  \
0         Lake Shore Dr & Monroe St            11.0   
1      Clinton St & Washington Blvd            31.0   
2      Sheffield Ave & Kingsbury St            15.0   
3           Carpenter St & Huron St            19.0   
4            Damen Ave & Pierce Ave            19.0   
...                             ...             ...   
50084         State St & Pearson St            27.0   
50085      Halsted St & 35th St (*)            16.0   
50086      Kingsbury St & Kinzie St            31.0   
50087          Clinton St & Lake St            23.0   
50088    Clarendon Ave & Leland Ave            15.0   

                     to_station_name  end_capacity  temperature  wind_speed  \
0              Michigan Ave & Oak St          15.0         73.9        12.7   
1               Wells St & Walton St          19.0         69.1         6.9   
2            Dearborn St & Monroe St          23.0         73.0        16.1   
3             Clark St & Randolph St          31.0         72.0        16.1   
4             Damen Ave & Pierce Ave          19.0         73.0        17.3   
...                              ...           ...          ...         ...   
50084              Clark St & Elm St          27.0          5.0        16.1   
50085            Union Ave & Root St          11.0          5.0        16.1   
50086  Halsted St & Blackhawk St (*)          20.0          5.0        16.1   
50087       Kingsbury St & Kinzie St          31.0          7.0        11.5   
50088     Clifton Ave & Lawrence Ave          15.0         10.9        15.0   

             events  
0      mostlycloudy  
1      partlycloudy  
2      mostlycloudy  
3      mostlycloudy  
4      partlycloudy  
...             ...  
50084  partlycloudy  
50085  partlycloudy  
50086  partlycloudy  
50087  partlycloudy  
50088  partlycloudy  

[50089 rows x 11 columns]

## What type of object is `bikes`?

Let's verify that `bikes` is indeed a DataFrame with the `type` function.

In [8]:
type(bikes)

pandas.core.frame.DataFrame

### Fully-qualified name

The above output is something called the **fully-qualified name**. Only the word after the last dot is the name of the type. We have now verified that the `bikes` variable has type `DataFrame`. 

The fully-qualified name always returns the package and module name of where the type was defined. The package name is the first part of the fully-qualified name and, in this case, is `pandas`. The module name is the word immediately preceding the name of the type. Here, it is `frame`.

### Package vs Module

A Python **package** is a directory containing other directories or modules that contain Python code. A Python **module** is a file (typically a text file ending in .py) that contains Python code. 

### Sub-packages

Any directory containing other directories or modules within a Python package is considered a **sub-package**. In this case, `core` is the sub-package.

### Where are the packages located on my machine?

Third-party packages are installed in the `site-packages` directory which itself is set up during Python installation. We can get the actual location with help from the standard library's `site` module's `getsitepackages` function.

In [9]:
import site
site.getsitepackages()

['c:\\Users\\nealv\\AppData\\Local\\Programs\\Python\\Python311',
 'c:\\Users\\nealv\\AppData\\Local\\Programs\\Python\\Python311\\Lib\\site-packages']

If you navigate to this directory in your file system, you'll find the 'pandas' directory. Within it will be a 'core' directory which will contain the 'frame.py' file. It is this file which contains Python code where the DataFrame class is defined.

## Select a single column from a DataFrame - a Series

To select a single column from a DataFrame, append a set of square brackets, `[]`, to the end of the DataFrame variable name. Place the column name as a string within those brackets to select it. This returns a single column of data as a pandas **Series**. This is a separate (but similar) type of object than a DataFrame.

Let's select the column name `tripduration`, assign it to a variable name, and output the first few values to the screen. The `head` and `tail` methods work the same as they do with DataFrames.

In [10]:
bikes.head(3)

gender            starttime             stoptime  tripduration  \
0   Male  2013-06-28 19:01:00  2013-06-28 19:17:00           993   
1   Male  2013-06-28 22:53:00  2013-06-28 23:03:00           623   
2   Male  2013-06-30 14:43:00  2013-06-30 15:01:00          1040   

              from_station_name  start_capacity          to_station_name  \
0     Lake Shore Dr & Monroe St            11.0    Michigan Ave & Oak St   
1  Clinton St & Washington Blvd            31.0     Wells St & Walton St   
2  Sheffield Ave & Kingsbury St            15.0  Dearborn St & Monroe St   

   end_capacity  temperature  wind_speed        events  
0          15.0         73.9        12.7  mostlycloudy  
1          19.0         69.1         6.9  partlycloudy  
2          23.0         73.0        16.1  mostlycloudy

In [12]:
td = bikes['tripduration']
td.head()

0     993
1     623
2    1040
3     667
4     130
Name: tripduration, dtype: int64

Select the last three values in the Series by passing the `tail` method the integer 3.

In [13]:
td.tail(3)

50086    824
50087    178
50088    214
Name: tripduration, dtype: int64

Let's verify that `td` has the type Series.

In [14]:
type(td)

pandas.core.series.Series

## Components of a Series

A Series is a similar type of object as a DataFrame but only contains a single dimension of data. It has two components - the **index** and the **data**. Let's take a look at a stylized Series graphic.

![0]

It's important to note that a Series has no rows and no columns. In appearance, it resembles a one-column DataFrame, but it technically has no columns. It just has a sequence of values that are labeled by an index.

### The index

A Series index serves as labels for the values. A single **label** or **name** always references a single value. In the above image, the index label **3** corresponds to the value 667. The Series index is virtually identical to the DataFrame index, so the same rules apply to it. Index values can be duplicated and can be types other than integers, such as strings. 

### Output of Series vs DataFrame

Notice that there is no nice HTML styling for the Series. It's just plain text. Below the Series display, you will see a few other items printed to the screen - the **name**, **length**, and **dtype**. These other items are NOT part of the Series itself and are just extra pieces of information to help you understand the Series.

* The **name** is not important right now. If the Series was formed from a column of a DataFrame, it will be set to that column name.
* The **length** is the number of values in the Series
* The **dtype** is the data type of the Series, which will be discussed in an upcoming chapter.

[0]: images/series_components.png

## Changing display options

pandas gives you the ability to change how the output on your screen is displayed. For instance, the default number of columns displayed for a DataFrame is 20, meaning that if your DataFrame has more than 20 columns, then only the first and last 10 columns will be shown on the screen. All the other columns will be hidden and unable to be displayed. This is problematic as many DataFrames have more than 20 columns.

### Get current option value with `get_option`

There are a few dozen display options you can control to change the visual representation of your DataFrame. It is not necessary to remember the option names as the official documentation provides descriptions for all [available options][1]. 

Let's first learn how to retrieve each option value with the `get_option` function. This is not a DataFrame method, but instead, a function that is accessed directly from `pd`.  Below are three of the most common options to change.

[1]: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html

In [15]:
pd.get_option('display.max_columns')

20

In [16]:
pd.get_option('display.max_rows')

60

In [17]:
pd.get_option('display.max_colwidth')

50

### Use the `set_option` function to change an option value

To change an option's value, use the `set_option` function. You can set as many options as you would like at one time. It's usage is a bit strange. Pass it the option name as a string and follow it immediately with the value you want to set it to. Continue this pattern of option name followed by new value to set as many options as you desire. Below, we set the maximum number of columns to 100 and the maximum number of rows to 4.

In [18]:
pd.set_option('display.max_columns', 100, 'display.max_rows', 4)

We now read in the housing dataset which contains 81 columns, all of which will be visible. Uncomment the lines to run them in your notebook.

In [ ]:
#housing = pd.read_csv('../data/housing.csv')
#housing

## Exercises

Use the `bikes` DataFrame for the following exercises.

### Exercise 1

<span  style="color:green; font-size:16px">Select the column `events`, the type of weather that was recorded, and assign it to a variable with the same name. Output the first 10 values of it.</span>

In [19]:
events = bikes['events']

events.head(10)

0    mostlycloudy
1    partlycloudy
         ...     
8          cloudy
9    mostlycloudy
Name: events, Length: 10, dtype: object

### Exercise 2

<span  style="color:green; font-size:16px">What type of object is `events`?</span>

In [20]:
type(events)

pandas.core.series.Series

### Exercise 3

<span  style="color:green; font-size:16px">Select the last two rows of the `bikes` DataFrame and assign it to the variable `bikes_last_2`. What type of object is `bikes_last_2`?</span>

In [21]:
bikes_last_2 = bikes.tail(2)

type(bikes_last_2)


pandas.core.frame.DataFrame

### Exercise 4

<span  style="color:green; font-size:16px">Use `pd.reset_option('all')` to reset the options to their default values. Test that this worked. </span>

In [22]:
pd.reset_option('all')

C:\Users\nealv\AppData\Local\Temp\ipykernel_192\2786130087.py:1: FutureWarning: data_manager option is deprecated and will be removed in a future version. Only the BlockManager will be available.
  pd.reset_option('all')
C:\Users\nealv\AppData\Local\Temp\ipykernel_192\2786130087.py:1: FutureWarning: use_inf_as_na option is deprecated and will be removed in a future version. Convert inf values to NaN before operating instead.
  pd.reset_option('all')
